### PR TITLE
(Feature) 288 cron task to delete fv records use existing cron

### PR DESCRIPTION
**Pull request by omerzam**
# Description

refactor for cron task to delete FV records using existing logic
 
About # ([288](https://app.zenhub.com/workspaces/gooddollar-5d50833888a83c6880d3e345/issues/gooddollar/goodserver/288))

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

# Checklist:
- [X] PR title matches follow: (Feature|Bug|Chore) Task Name
- [X] My code follows the style guidelines of this project
- [X] I have followed all the instructions described in the initial task (check Definitions of Done)
- [X] I have performed a self-review of my own code
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I have added reference to a related issue in the repository
- [ ] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request (for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [ ] @mentions of the person or team responsible for reviewing proposed changes

### CODE REVIEW

**Review by sirpy — CHANGES_REQUESTED** [review]
add relevant indexes for disposing queue so queries are executed efficiently
probably a composed index on {createdDate, subject.executeAt}
go over queries and make sure they have the relevant indexes

File: src/server/cron/taskUtil.js [hunk 1]
```diff
@@ -2,6 +2,9 @@
 export const DISPOSE_ENROLLMENTS_TASK = 'verification/dispose_enrollments'
 
 export const scheduleDisposalTask = async (storage, enrollmentIdentifier, executeAt): Promise<DelayedTaskRecord> => {
-  await storage.cancelTasksQueued(DISPOSE_ENROLLMENTS_TASK, { 'subject.enrollmentIdentifier': enrollmentIdentifier })
+  await storage.cancelTasksQueued(DISPOSE_ENROLLMENTS_TASK, {
+    'subject.enrollmentIdentifier': enrollmentIdentifier,
+    'subject.executeAt': executeAt
```
Comment (sirpy) [inline src/server/cron/taskUtil.js:7]: maybe match to same input format as enqueueTask ie without the subject prefix

**Review by sirpy — CHANGES_REQUESTED** [review]

File: src/server/verification/__tests__/verificationAPI.js [hunk 2]
```diff
@@ -145,7 +147,7 @@ describe('verificationAPI', () => {
 
     beforeEach(async () => {
       await storage.updateUser({ identifier: userIdentifier, isVerified: false, claimQueue: null })
-      await storage.taskModel.deleteMany({ subject: enrollmentIdentifier })
+      await storage.taskModel.deleteMany({})
```
Comment (serdiukov-o-nordwhale) [inline src/server/verification/__tests__/verificationAPI.js:150]:
```suggestion
      await storage.taskModel.deleteMany({ "subject.enrollmentIdentifier": enrollmentIdentifier })
```

File: src/server/cron/taskUtil.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+//@flow
+export const DISPOSE_ENROLLMENTS_TASK = 'verification/dispose_enrollments'
+
+export const scheduleDisposalTask = async (storage, enrollmentIdentifier, executeAt): Promise<DelayedTaskRecord> => {
+  await storage.cancelTasksQueued(DISPOSE_ENROLLMENTS_TASK, {
+    'subject.enrollmentIdentifier': enrollmentIdentifier,
+    'subject.executeAt': executeAt
+  })
+  return await storage.enqueueTask(DISPOSE_ENROLLMENTS_TASK, { enrollmentIdentifier, executeAt })
```
Comment (serdiukov-o-nordwhale) [inline src/server/cron/taskUtil.js:9]:
```suggestion

  return storage.enqueueTask(DISPOSE_ENROLLMENTS_TASK, { enrollmentIdentifier, executeAt })
```

File: src/server/cron/taskUtil.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+//@flow
+export const DISPOSE_ENROLLMENTS_TASK = 'verification/dispose_enrollments'
+
+export const scheduleDisposalTask = async (storage, enrollmentIdentifier, executeAt): Promise<DelayedTaskRecord> => {
```
Comment (serdiukov-o-nordwhale) [inline src/server/cron/taskUtil.js:4]:
```suggestion
// eslint-disable-next-line require-await
export const scheduleDisposalTask = async (storage, enrollmentIdentifier, executeAt): Promise<DelayedTaskRecord> => {
```

File: src/server/cron/taskUtil.js (hunk 3 above)
Comment (serdiukov-o-nordwhale) [reply src/server/cron/taskUtil.js:9]: we're using `return await` where's exactly needed (e.g. inside try ... catch}

File: src/server/db/mongo/user-privat-provider.js [hunk 5]
```diff
@@ -232,7 +232,17 @@ class UserPrivate {
...
 
   /**
-   * Fetches tasks of the type specifid with optional filtering and locks them by setting running status
+   * cancel queued task
+   * @param {string} taskName
+   * @param {object} filters
+   */
+  async cancelTasksQueued(taskName: string, filters: object = {}): Promise<void> {
+    const { taskModel } = this
+    await taskModel.deleteOne({ ...filters, taskName })
```
Comment (serdiukov-o-nordwhale) [inline src/server/db/mongo/user-privat-provider.js:241]:
```suggestion

    await taskModel.deleteOne({ ...filters, taskName })
```

File: src/server/verification/__tests__/verificationAPI.js [hunk 6]
```diff
@@ -223,7 +225,7 @@ describe('verificationAPI', () => {
     })
 
     test('PUT /verify/face/:enrollmentIdentifier returns 400 if user is being deleted', async () => {
-      await storage.enqueueTask(DISPOSE_ENROLLMENTS_TASK, enrollmentIdentifier)
+      await storage.enqueueTask(DISPOSE_ENROLLMENTS_TASK, { enrollmentIdentifier, executeAt: 'account-removal' })
```
Comment (serdiukov-o-nordwhale) [inline src/server/verification/__tests__/verificationAPI.js:228]: pls use taskUtil for that

File: src/server/verification/processor/EnrollmentProcessor.js [hunk 7]
```diff
@@ -82,15 +82,15 @@ class EnrollmentProcessor {
...
+        subject: { enrollmentIdentifier, executeAt: 'account-removal' }
       })
 
       log.info('Got disposal state for enrollment', { enrollmentIdentifier, isDisposing })
       return isDisposing
     } catch (exception) {
       const error = exception.message
 
-      log.warn("Coundn't check disposal state for enrollment", { enrollmentIdentifier, error })
+      log.warn("Couldn't check disposal state for enrollment", { enrollmentIdentifier, error })
```
Comment (serdiukov-o-nordwhale) [inline src/server/verification/processor/EnrollmentProcessor.js:93]: tnx for the typos fixes

**Review by serdiukov-o-nordwhale — APPROVED** [review]
Good job, thanks!

File: src/server/cron/taskUtil.js (hunk 1 above)
Comment (serdiukov-o-nordwhale) [inline src/server/cron/taskUtil.js:7]:
```suggestion
```

no, we actually need to remove **any** task for that enrollmentIdentifier before add the new one. so we have to check  'subject.enrollmentIdentifier' only in the filters

Comment (serdiukov-o-nordwhale) [reply src/server/cron/taskUtil.js:7]:
```suggestion
'subject.enrollmentIdentifier': enrollmentIdentifier
```

no, we actually need to remove **any** task for that enrollmentIdentifier before add the new one. so we have to check  'subject.enrollmentIdentifier' only in the filters

Comment (omerzam) [reply src/server/cron/taskUtil.js:7]: that would mean having to add an index for that query FYI

File: src/server/db/mongo/user-privat-provider.js (hunk 5 above)
Comment (omerzam) [reply src/server/db/mongo/user-privat-provider.js:241]: is it just me or did you change nothing in this suggestion? am I missing something?

Comment (serdiukov-o-nordwhale) [reply src/server/db/mongo/user-privat-provider.js:241]: is suggested to addd just  an empty line before (to make code more readable)

File: src/server/verification/processor/EnrollmentProcessor.js (hunk 7 above)
Comment (omerzam) [reply src/server/verification/processor/EnrollmentProcessor.js:93]: you're welcome
check this out: 
https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker

File: src/server/cron/taskUtil.js (hunk 1 above)
Comment (sirpy) [reply src/server/cron/taskUtil.js]: @omerzam ok. make sure we have the correct index.

**Review by sirpy — APPROVED** [review]
@serdiukov-o-nordwhale solve conflicts, merge, move to QA

**Comment by serdiukov-o-nordwhale** [comment]
@sirpy i resolved all conflicts and adjusted code & tests
**but** it's not tested yet, _will test tomorrow_ then it could be merged

**Pull request merged**